README.md
# Blog

[![Build Status](https://github.com/kuro-kuroite/kuroite/workflows/build/badge.svg)](https://github.com/kuro-kuroite/kuroite/actions)

<!-- textlint-disable ja-technical-writing/sentence-length -->

[![GitHub License](https://img.shields.io/badge/license-MIT-blue.svg)](./LICENSE.md)
[![GitHub release (latest SemVer)](https://img.shields.io/github/v/release/kuro-kuroite/kuroite)](https://github.com/kuro-kuroite/kuroite/releases)
[![Coverage Status](https://codecov.io/gh/kuro-kuroite/kuroite/graph/badge.svg?branch=main)](https://codecov.io/gh/kuro-kuroite/kuroite)
[![dependency Status](https://img.shields.io/david/kuro-kuroite/kuroite.svg?maxAge=1000)](https://david-dm.org/kuro-kuroite/kuroite)
[![devDependency Status](https://img.shields.io/david/dev/kuro-kuroite/kuroite.svg?maxAge=1000)](https://david-dm.org/kuro-kuroite/kuroite)
[![GitHub top language](https://img.shields.io/github/languages/top/kuro-kuroite/kuroite)](https://github.com/kuro-kuroite/kuroite/search?l=typescript)
[![Renovate enabled](https://img.shields.io/badge/renovate-enabled-brightgreen.svg)](https://renovatebot.com/)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

<!-- textlint-enable -->

[![Storybook](https://raw.githubusercontent.com/storybooks/brand/master/badge/badge-storybook.svg)](https://storybook.js.org/)
[![tested with jest](https://img.shields.io/badge/tested_with-jest-99424f.svg)](https://github.com/facebook/jest)
[![tested with Cypress](https://img.shields.io/badge/tested_with-Cypress-04C38E.svg)](https://www.cypress.io/)

[![Conventional Commits](https://img.shields.io/badge/Conventional_Commits-friendly-brightgreen.svg)](https://conventionalcommits.org)
[![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)
[![code style prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg)](https://github.com/prettier/prettier)

# 概要

モダンな Web フロントエンドの技術を中心に発信する技術ブログ。

[https://blog.kuroite.dev](https://blog.kuroite.dev)

<details>
  <summary>
    記事ページのデモ画像
  </summary>
  <div style="display: flex;">

[<figure style="display: inline-flex; flex-direction: column;"><img src="./docs/assets/blogPost_mobile.gif" width="100px" height="200px" alt="blog post page in mobile" /><figcaption>(モバイル版)</figcaption></figure>](./docs/assets/blogPost_mobile.gif)
[<figure style="display: inline-flex; flex-direction: column;"><img src="./docs/assets/blogPost_pc.gif" width="400px" height="200px" alt="blog post page in pc" /><figcaption>(pc 版)</figcaption></figure>](./docs/assets/blogPost_pc.gif)

  </div>
</details>

## 環境

<!-- TODO: システム構成図 -->

## 機能一覧

以下の技術を使い、このブログを構成している。

- 静的サイトジェネレーター for Jamstack (Gatsby.js)
- モバイルファースト
- ダークモード対応
- 目次リンク
- Twitter、Hatena Bookmark 共有リンク
- プログレスバー
- 「ラベルの翻訳 [^1]」の英語対応
- コードブロック
  - ラベル表示（ファイル名）
  - シンタックスハイライト
- コンポーネントのアクセシビリティテスト
- RSS 対応
- PWA 対応
- ページ
  - 記事ページ
  - 記事一覧ページ
  - タグの条件付き記事一覧ページ
  - 自己紹介ページ

### ブログ構成に必要な機能

- プライバシーポリシー
- robots.txt
- canonical
- meta tag
- Google Analytics

#### aside: 注意点

本ブログは、Internet Explorer に「**非**」対応である。
これは、開発元の Microsoft が今後機能追加をせずサポートを打ち切るため（[Microsoft 365 アプリの IE 11 サポート終了](support-ie11-url)）。
そして、サイズを小さくしブログの読み込み時間を短くするためである。

[support-ie11-url]: https://techcommunity.microsoft.com/t5/microsoft-365-blog/microsoft-365-apps-say-farewell-to-internet-explorer-11-and/ba-p/1591666

## 技術一覧

[<img src="./docs/assets/typescript.svg" width="16px" height="16px" alt="TypeScript icon" />](https://www.typescriptlang.org/ 'TypeScript: Typed JavaScript at Any Scale.')
[<img src="./docs/assets/react.svg" width="16px" height="16px" alt="React icon" />](https://reactjs.org/ 'React – A JavaScript library for building user interfaces')
[<img src="./docs/assets/eslint.svg" width="16px" height="16px" alt="ESLint icon" />](https://eslint.org/ 'ESLint - Pluggable JavaScript linter')
[<img src="./docs/assets/prettier.svg" width="16px" height="16px" alt="Prettier icon" />](https://prettier.io/ 'Prettier · Opinionated Code Formatter')
[<img src="./docs/assets/redux_toolkit.svg" width="16px" height="16px" alt="Redux Toolkit icon" />](https://redux-toolkit.js.org/ 'Redux Toolkit | Redux Toolkit')
[<img src="./docs/assets/jest.svg" width="16px" height="16px" alt="Jest icon" />](https://jestjs.io/ 'Jest · 🃏 Delightful JavaScript Testing')
[<img src="./docs/assets/storybook.svg" width="16px" height="16px" alt="Storybook icon" />](https://storybook.js.org/ 'Storybook: UI component explorer for frontend developers')
[<img src="./docs/assets/cypress.svg" width="16px" height="16px" alt="Cypress icon" />](https://www.cypress.io/ 'JavaScript End to End Testing Framework | cypress.io')
[<img src="./docs/assets/gatsby.svg" width="16px" height="16px" alt="Gatsby.js icon" />](https://www.gatsbyjs.com/ 'Gatsby')
[<img src="./docs/assets/postcss.svg" width="16px" height="16px" alt="PostCSS icon" />](https://postcss.org/ 'PostCSS - a tool for transforming CSS with JavaScript')
[<img src="./docs/assets/autoprefixer.svg" width="16px" height="16px" alt="Autoprefixer icon" />](https://autoprefixer.github.io/ 'Autoprefixer CSS online')
[<img src="./docs/assets/stylelint.svg" width="16px" height="16px" alt="stylelint icon" />](https://stylelint.io/ 'stylelint.io')
[<img src="./docs/assets/tailwindcss.svg" width="16px" height="16px" alt="Tailwind CSS icon" />](https://tailwindcss.com/ 'Tailwind CSS - A Utility-First CSS Framework for Rapidly Building Custom Designs')
[<img src="./docs/assets/css_modules.svg" width="16px" height="16px" alt="CSS Modules icon" />](https://github.com/css-modules/css-modules/blob/master/README.md 'css-modules/README.md at master · css-modules/css-modules')
[<img src="./docs/assets/commitlint.svg" width="16px" height="16px" alt="commitlint icon" />](https://commitlint.js.org/#/ 'commitlint - Lint commit messages')
[<img src="./docs/assets/textlint.svg" width="16px" height="16px" alt="textlint icon" />](https://textlint.github.io/ 'textlint · The pluggable linting tool for text and markdown')

[<img src="./docs/assets/github_actions.png" width="16px" height="16px" alt="GitHub Actions icon" />](https://github.com/features/actions 'Features • GitHub Actions')
[<img src="./docs/assets/docker.png" width="16px" height="16px" alt="Docker icon" />](https://www.docker.com/ 'Empowering App Development for Developers | Docker')
[<img src="./docs/assets/renovate.png" width="16px" height="16px" alt="Renovate icon" />](https://docs.renovatebot.com/ 'Renovate Docs | Renovate Docs')
[<img src="./docs/assets/semantic_release.svg" width="16px" height="16px" alt="semantic release icon" />](https://semantic-release.gitbook.io/semantic-release/ 'Introduction - semantic-release')
[<img src="./docs/assets/netlify.svg" width="16px" height="16px" alt="Netlify icon" />](https://www.netlify.com/ 'Netlify: All-in-one platform for automating modern web projects')

詳細は、[技術選定.md](./docs/技術選定.md '技術選定.md') で紹介。

# 開発者向け

## 事前要件

## インストール

### `develop`

```zsh
# pre
cd path/to/blog
git clone https://github.com/kuro-kuroite/kuroite.git

  # plan A: if you use VSCode, and `ms-vscode-remote.remote-containers` extension
  code path/to/blog/kuroite
  mv .env.local.example .env.local
  vim .env.local # for Docker
  ## open `Docker.app` or Docker something in Spotlight(Command + Space > `Docker.app`)
  ## run `Remote-Containers: Reopen in Container` in Command Palette(Command + Shift + P)
  ## open Terminal in VSCode(Control + `)
  ### click `Install` to install extensions from right bottom popup of `Do you want to install the recommended extensions for docker-compose.yml?`

  # plan B: otherwise
  cd path/to/blog/kuroite
  mv .env.local.example .env.local
  vim .env.local # for Docker
  docker-compose build --no-cache # only first time
  docker-compose up -d
  docker-compose exec web zsh
  zsh ./config/docker/devcontainer/post-create-command.sh
  ## if you use VSCode
  ### click `Install` to install extensions from right bottom popup of `Do you want to install the recommended extensions for this repository?`

mv .env.development.example .env.development
vim .env.development # for `npm run develop`
mv .env.production.example .env.production
vim .env.production # for `npm run build && npm run serve`

# `develop`
pwd # /code
npx gatsby develop -H=$HOST # or `npm run develop`
# open http://localhost:8000 in browser

# post
  # plan A
  ## click `Dev Container: React` > `Close Remote Connection` from left bottom green button

  # plan B
  exit
  docker-compose down # or `docker-compose down --volumes`

# click `Quit Docker Desktop` or Docker something
```

### `deploy`

```zsh
# pre
git branch feature/something
git switch feature/something

npx gatsby build # or `npm run build`
npx gatsby serve -H=$HOST # or `npm run serve`
# check http://localhost:9000 in browser

# `git add` and `git commit`
git push origin feature/something

# make Pull Request for main in https://github.com/kuro-kuroite/kuroite/compare

# `deploy`
# main に merge すると、CI(build) が build し、Netlify に deploy する
```

### License

[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](./LICENSE.md)

### 脚注

[^1]: [Kaizen Ad における SPA での i18n への取り組みと手法について - Kaizen Platform 開発者ブログ](https://developer.kaizenplatform.com/entry/axross/2018-07-02 'Kaizen AdにおけるSPAでのi18nへの取り組みと手法について - Kaizen Platform 開発者ブログ')


CHANGELOG.md
## [1.4.0](https://github.com/kuro-kuroite/test-kuroite/compare/v1.3.0...v1.4.0) (2021-04-09)


### Features

* **privacy-policy:** :sparkles: add privacyPolicy with TSX, PostCSS, Tailwind, and I18n ([12006cb](https://github.com/kuro-kuroite/test-kuroite/commit/12006cb998576814ff43b3ce3bc82f34b7030189))

## [1.3.0](https://github.com/kuro-kuroite/test-kuroite/compare/v1.2.0...v1.3.0) (2021-04-09)


### Features

* **tag:** :sparkles: add tag with TSX, PostCSS, Tailwind, and I18n ([1479de6](https://github.com/kuro-kuroite/test-kuroite/commit/1479de62154a7bf7f6990481197fce52d5510372))

## [1.2.0](https://github.com/kuro-kuroite/test-kuroite/compare/v1.1.0...v1.2.0) (2021-04-09)


### Features

* **about:** :sparkles: add about with TSX, PostCSS, Tailwind, and I18n ([11cca6f](https://github.com/kuro-kuroite/test-kuroite/commit/11cca6f88c3da86f4e555a5df6e11a53b0752cd2))

## [1.1.0](https://github.com/kuro-kuroite/test-kuroite/compare/v1.0.0...v1.1.0) (2021-04-09)


### Features

* **index:** :sparkles: add index with TSX, PostCSS, Tailwind, and I18n ([4288b98](https://github.com/kuro-kuroite/test-kuroite/commit/4288b98f5fb36332a9ebbf275aab48b7e3efc4c3))

## 1.0.0 (2021-04-09)


### Features

* **remark,blogpost:** :sparkles: add blogPost with remark, TSX, PostCSS, Tailwind, and I18n ([0f5f618](https://github.com/kuro-kuroite/test-kuroite/commit/0f5f61855dbee298c2d524c3ae2e0d212dbca273))

## 1.0.0 (2021-04-01)


### Features

* **remark,blogpost:** :sparkles: add blogPost with remark, TSX, PostCSS, Tailwind, and I18n ([8c78991](https://github.com/kuro-kuroite/test-kuroite/commit/8c78991c787f7457a7da47561b53ffc3b024c327))


cypress/integration/404.spec.ts
import { getByTestId } from '../support/getByTestId';

describe('/404', () => {
  beforeEach(() => {
    cy.visit('/404');
    cy.injectAxe();
  });

  specify(
    'ロードした時に，dark modeとlight modeのアクセシビリティを確認する',
    () => {
      // HACK: checkA11yの前に任意のテストを実行し，DOMを描画
      cy.get('main').should('have.length', 1);

      getByTestId('darkModeCheckbox').click({ force: true });
      cy.checkA11y();

      getByTestId('darkModeCheckbox').click({ force: true });
      cy.checkA11y();
    }
  );

  afterEach(() => {
    cy.document().its('body').toMatchSnapshot();
  });
});


cypress/integration/sample.spec.ts
import { getByTestId } from '../support/getByTestId';

describe('/blog/sample/', () => {
  beforeEach(() => {
    cy.visit('/blog/sample');
    cy.injectAxe();
  });

  specify(
    'ロードした時に，dark modeとlight modeのアクセシビリティを確認する',
    () => {
      // HACK: checkA11yの前に任意のテストを実行し，DOMを描画
      cy.get('main').should('have.length', 1);

      getByTestId('darkModeCheckbox').click({ force: true });
      cy.checkA11y();

      getByTestId('darkModeCheckbox').click({ force: true });
      cy.checkA11y();
    }
  );

  afterEach(() => {
    cy.document().its('body').toMatchSnapshot();
  });
});


config/gatsby-node.ts
import { GatsbyNode } from 'gatsby';
import { createFilePath } from 'gatsby-source-filesystem';
import path from 'path';

export const createPages: GatsbyNode['createPages'] = async ({
  actions,
  graphql,
  reporter,
}) => {
  // eslint-disable-next-line @typescript-eslint/unbound-method
  const { createPage } = actions;
  // HACK: src/hack/gatsby-plugin-typegen/graphql-type-pages.tsのuseDummy()と対応
  const result = await graphql<GatsbyTypes.PagesQuery>(`
    query Pages {
      posts: allMarkdownRemark(
        filter: {
          frontmatter: { status: { ne: "draft" } }
          fileAbsolutePath: { regex: "/content/blog/" }
        }
        sort: {
          order: DESC
          fields: [frontmatter___createdAt, frontmatter___updatedAt]
        }
        limit: 1000
      ) {
        edges {
          next {
            frontmatter {
              title
            }
            fields {
              slug
            }
          }
          previous {
            fields {
              slug
            }
            frontmatter {
              title
            }
          }
          node {
            fields {
              slug
            }
            frontmatter {
              title
            }
          }
        }
      }
    }
  `);

  if (result.errors) {
    reporter.panicOnBuild(result.errors);

    return;
  }

  // NOTE: for blogPost.tsx
  if (!result.data?.posts?.edges) {
    return;
  }

  const posts = result.data.posts.edges;
  const blogPostTemplate = path.resolve(`./src/templates/blogPost.tsx`);

  posts.forEach(({ next, node, previous }) => {
    if (!node.fields?.slug) {
      reporter.panicOnBuild('no slug');

      return;
    }

    createPage({
      component: blogPostTemplate,
      context: {
        next: next,
        previous: previous,
        slug: node.fields.slug,
      },
      path: node.fields.slug,
    });
  });
};

// FYI: https://qiita.com/AumyF/items/a6a8400cf9f5f2ce488f#configgatsby-nodets%E3%82%92%E6%9B%B8%E3%81%8F
export const onCreateNode: GatsbyNode['onCreateNode'] = ({
  actions,
  getNode,
  node,
}) => {
  // eslint-disable-next-line @typescript-eslint/unbound-method
  const { createNodeField } = actions;

  if (node.internal.type === 'MarkdownRemark') {
    const value = createFilePath({ getNode, node });
    createNodeField({
      name: 'slug',
      node,
      value,
    });
  }
};


config/gatsby-config.ts
import { config } from 'dotenv';
import { GatsbyConfig } from 'gatsby';
import path from 'path';

import { colors } from './colors.config';

// FYI: https://www.gatsbyjs.com/docs/how-to/local-development/environment-variables/#google-analytics-env-var-example
const activeEnv =
  process.env.GATSBY_ACTIVE_ENV || process.env.NODE_ENV || 'development';
console.log(`Using environment config: '${activeEnv}'`);
config({
  path: `.env.${activeEnv}`,
});

const title = process.env.SITE_TITLE ?? '';
const siteUrl = process.env.SITE_URL ?? '';
const gatsbyConfig: GatsbyConfig = {
  plugins: [
    {
      options: {
        gtagConfig: {
          anonymize_ip: true,
          cookie_expires: 0,
        },
        pluginConfig: {
          head: false,
          respectDNT: true,
        },
        trackingIds: [process.env.GA_TRACKING_ID],
      },
      resolve: `gatsby-plugin-google-gtag`,
    },
    'gatsby-plugin-postcss',
    {
      options: {
        outputPath: './typings/graphql-type/index.d.ts',
      },
      resolve: 'gatsby-plugin-typegen',
    },
    {
      options: {
        showSpinner: false,
      },
      resolve: `gatsby-plugin-nprogress`,
    },
    {
      options: {
        name: 'blog',
        path: path.resolve('./content/blog'),
      },
      resolve: 'gatsby-source-filesystem',
    },
    {
      options: {
        name: 'assets',
        path: path.resolve('./content/assets'),
      },
      resolve: 'gatsby-source-filesystem',
    },
    {
      options: {
        footnotes: true,
        gfm: true,
        plugins: [
          {
            options: {},
            resolve: 'gatsby-remark-code-titles',
          },
          {
            options: {
              className: 'headerAnchor',
              elements: ['h1', 'h2', 'h3'],
              maintainCase: false,
              // NOTE: ジャンプ時の上余白 (base.cssで設定: scroll-padding-top)
              offsetY: 80,
            },
            resolve: 'gatsby-remark-autolink-headers',
          },
          {
            options: {
              theme: 'dark-plus',
            },
            resolve: 'gatsby-remark-shiki',
          },
          {
            options: {
              wrapperStyle: 'margin-bottom: 1.0725rem',
            },
            resolve: 'gatsby-remark-responsive-iframe',
          },
          'gatsby-remark-copy-linked-files',
          'gatsby-remark-smartypants',
          {
            options: {
              maxWidth: 590,
            },
            resolve: 'gatsby-remark-images',
          },
        ],
      },
      resolve: 'gatsby-transformer-remark',
    },
    'gatsby-transformer-sharp',
    'gatsby-plugin-sharp',
    {
      options: {
        //trackingId: 'ADD YOUR TRACKING ID HERE',
      },
      resolve: 'gatsby-plugin-google-analytics',
    },
    {
      options: {
        feeds: [
          {
            output: '/rss.xml',
            // HACK: src/hack/gatsby-plugin-typegen/graphql-type-rss.tsのuseDummy()と対応
            query: `
              query Rss {
                allMarkdownRemark(
                  sort: {
                    order: DESC
                    fields: [frontmatter___createdAt, frontmatter___updatedAt]
                  }
                  filter: {
                    fileAbsolutePath: { glob: "**/content/blog/**/*.md" }
                    frontmatter: { status: { ne: "draft" } }
                  }
                  limit: 1000
                ) {
                  edges {
                    node {
                      excerpt
                      html
                      fields {
                        slug
                      }
                      frontmatter {
                        title
                        updatedAt
                        createdAt
                        description
                      }
                    }
                  }
                }
              }
            `,
            // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
            serialize: ({
              query: { allMarkdownRemark, site },
            }: {
              query: {
                allMarkdownRemark: GatsbyTypes.RssQuery['allMarkdownRemark'];
                site: GatsbyTypes.RssSiteQuery['site'];
              };
            }) => {
              return allMarkdownRemark.edges.map(({ node }) => {
                if (!site?.siteMetadata?.siteUrl) {
                  throw new Error('no siteUrl');
                }

                if (!node) {
                  throw new Error('no node');
                }

                if (!node?.html) {
                  throw new Error('no html');
                }

                if (!node?.fields?.slug) {
                  throw new Error('no slug');
                }

                if (!node?.frontmatter) {
                  throw new Error('no frontmatter');
                }

                if (!node?.frontmatter?.createdAt) {
                  throw new Error('no createdAt');
                }

                if (!node?.frontmatter?.description) {
                  throw new Error('no description');
                }

                if (!node?.frontmatter?.title) {
                  throw new Error('no title');
                }

                if (!node?.frontmatter?.updatedAt) {
                  throw new Error('no updatedAt');
                }

                return Object.assign({}, node.frontmatter, {
                  custom_elements: [{ 'content:encoded': node.html }],
                  date: node.frontmatter.createdAt,
                  description: node.frontmatter.description,
                  guid: `${site.siteMetadata.siteUrl}${node.fields.slug}`,
                  url: `${site.siteMetadata.siteUrl}${node.fields.slug}`,
                });
              });
            },
            title: `RSS Feed | ${title}`,
          },
        ],
        // HACK: src/hack/gatsby-plugin-typegen/graphql-type-rss-site.tsのuseDummy()と対応
        query: `
          query RssSite {
            site {
              siteMetadata {
                title
                description
                siteUrl
                site_url: siteUrl
              }
            }
          }
        `,
      },
      resolve: 'gatsby-plugin-feed',
    },
    {
      options: {
        background_color: colors.black[800],
        display: 'standalone',
        icon: path.resolve('./src/components/Layout/LogoLink/assets/logo.svg'),
        icon_options: {
          purpose: 'any maskable',
        },
        name: 'Starter',
        short_name: 'Starter',
        start_url: '/',
        theme_color: colors.orange.DEFAULT,
      },
      resolve: 'gatsby-plugin-manifest',
    },
    'gatsby-plugin-offline',
    'gatsby-plugin-react-helmet',
    'gatsby-plugin-sitemap',
    'gatsby-plugin-robots-txt',
    {
      options: {
        siteUrl: siteUrl,
        stripQueryString: true,
      },
      resolve: 'gatsby-plugin-canonical-urls',
    },
  ],
  siteMetadata: {
    author: {
      name: process.env.SITE_AUTHOR_NAME,
      summary: process.env.SITE_AUTHOR_SUMMARY,
    },
    description: process.env.SITE_DESCRIPTION,
    siteUrl: siteUrl,
    social: {
      gitHub: process.env.ACCOUNT_GITHUB,
      twitter: process.env.ACCOUNT_TWITTER,
    },
    title: title,
  },
};

export default gatsbyConfig;
